src/client/runner/automation/util.js

```diff
@@ -11,27 +12,100 @@ var positionUtils   = testCafeCore.positionUtils;
 var styleUtils      = testCafeCore.styleUtils;
 
 
-function focusElementByLabel (label, callback) {
-    var doc              = domUtils.findDocument(label);
-    var focusableElement = doc.getElementById(label.getAttribute('for'));
+function setCaretPosition (element, caretPos) {
+    var isTextEditable    = domUtils.isTextEditableElement(element);
+    var isContentEditable = domUtils.isContentEditableElement(element);
 
-    if (focusableElement && domUtils.getActiveElement() !== focusableElement)
-        focusBlurSandbox.focus(focusableElement, callback, false, true);
-    else if (callback)
-        callback();
+    if (isTextEditable || isContentEditable) {
+        if (isContentEditable && isNaN(parseInt(caretPos, 10)))
+            textSelection.setCursorToLastVisiblePosition(element);
+        else {
+            var position = isNaN(parseInt(caretPos, 10)) ? element.value.length : caretPos;
+
+            textSelection.select(element, position, position);
+        }
+    }
+    else {
+        // NOTE: if focus is called for not contentEditable element (like 'img' or 'button') inside
+        // contentEditable parent we should try to set right window selection. Generally we can't
+        // set right window selection object because after selection setup window.getSelection method
+        // returns  a different object depending on the browser.
+        var contentEditableParent = contentEditable.findContentEditableParent(element);
+
+        if (contentEditableParent)
+            textSelection.setCursorToLastVisiblePosition(contentEditable.findContentEditableParent(contentEditableParent));
+    }
+}
+
+export function focusAndSetSelection (element, simulateFocus, caretPos) {
+    return new Promise(resolve => {
+        var activeElement               = domUtils.getActiveElement();
+        var isTextEditable              = domUtils.isTextEditableElement(element);
+        var labelWithForAttr            = domUtils.closest(element, 'label[for]');
+        var shouldFocusByRelatedElement = !domUtils.isElementFocusable(element) && labelWithForAttr;
+        var isContentEditable           = domUtils.isContentEditableElement(element);
+        var elementForFocus             = isContentEditable ? contentEditable.findContentEditableParent(element) : element;
+
+        // NOTE: in WebKit if selection was never set in an input element focus method selects all text
+        // of this element. So we should select call select before focus to set caret to first symbol.
+        if (simulateFocus && browserUtils.isWebKit && isTextEditable)
+            textSelection.select(element, 0, 0);
+
+        // NOTE: we should call focus for element related with 'label' with attribute 'for'
+        if (shouldFocusByRelatedElement) {
+            if (simulateFocus)
+                focusByRelatedElement(labelWithForAttr);
+
+            return resolve();
```

Comment:
I think we should have the same type of return value. I'll separate them, thanks for notice.